### DISCUSSION TITLE
The combined power of F# and C#

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#73] [General]
# The combined power of F# and C#

Where C# is the most dominant language in the .NET world, other languages are built on top of the Framework that deserves their respective place. F# is strong when it comes down to functional programming! In this blog post, we will leverage the power of F# and C# to showcase where both excel!

https://steven-giesel.com/blogPost/2f70d926-ec92-4dfe-b278-18f78078253d

**Comment by aweadd014w** [comment]
hi Steven! Thank you for sharing! I can't find the source codes in the repository. Could you please have a look?

**Reply by linkdotnet** [reply]
Yes - because I forgot to push the changes :)
That is on me, sorry. Will do this evening.

**Comment by saint4eva** [comment]
Kindly use the required keyword and remove the constructor. And if you like, you can use records also

**Reply by linkdotnet** [reply]
Of course, modern language features would apply here. Thanks for the input - for the sake of simplicity, I left them out as I did not need them.

**Reply by saint4eva** [reply]
It would bring terseness and simplicity to the codebase. I appreciate and read your posts always. Keep up the good work.

**Reply by linkdotnet** [reply]
@saint4eva thanks for the kind words - I did follow your recommendation and updated the blog post to use more modern language features like `record` to make the C# more concise.

**Reply by ubik** [reply]
Product, OrderLine & Discount should be records too. C# pattern matching could be used in CalculateTotalPrice.
With these modifications C# code will look very similar to F# but slighty clearer in my opinion.

```
public record Product(string Name, decimal Price);
public record OrderLine(Product Product, int Quantity);

public enum DiscountType { Percentage, FixedAmount };
public record Discount(DiscountType Type, decimal Value);

public record Order(List<OrderLine> OrderLines, Discount Discount)
{
    public decimal CalculateTotalPrice() 
    {
        var subtotal = OrderLines.Sum(ol => ol.Product.Price * ol.Quantity);
        var totalDiscount = Discount switch
        {
            { Type: DiscountType.Percentage } => subtotal * Discount.Value / 100M,
            { Type: DiscountType.FixedAmount } => Discount.Value,
            _ => 0M
        };

        return subtotal - totalDiscount;
    }
}
```

**Reply by linkdotnet** [reply]
Fair point - it would be a bit less verbose with records and more compact. Things like discriminated unions still don't work with C#, which is a pity.

**Reply by keithn** [reply]
Discriminated Unions don't work in C#? yeah they do

```C#
public record Product(string Name, decimal Price);
public record OrderLine(Product Product, int Quantity);

public abstract record Discount 
{
    public record Percentage(decimal Value) : Discount();
    public record FixedAmount(decimal Value) : Discount();
}

public record Order(List<OrderLine> OrderLines, Discount Discount)
{
    public decimal CalculateTotalPrice()
    {
        var subtotal = OrderLines.Sum(ol => ol.Product.Price * ol.Quantity);
        return subtotal - Discount switch
        {
            Discount.Percentage { } percentage => subtotal * percentage.Value / 100M,
            Discount.FixedAmount { } amount => amount.Value,
            _ => 0M
        };
    }
}
```

**Reply by linkdotnet** [reply]
That is pattern matching and only has a small overlap with discriminated unions (that does not need inheritance).

**Reply by keithn** [reply]
this is how the C# proposal for DUs will implement them, what makes you think it isn't?  F# does it the same way essentially.  A discriminated type + pattern matching.  The extra it provides is exhaustive checking, but that's more of an add on in the philosophy of functional programming ( I think it may end up as a warning in C#).

**Reply by keithn** [reply]
here's an example I did ages ago based on F# for fun and profits "sell" for DUs.... https://fsharpforfunandprofit.com/posts/designing-for-correctness/    , and in C#  https://gist.github.com/keithn/ceeeed5f7eb567e1b2333747065d10dd      If you are having trouble, just ignore the inheritance and you'll see it's just the same.

**Reply by linkdotnet** [reply]
Fair point - I took your code @keithn and if you [convert from F# to C#](https://sharplab.io/#v2:DYLgZgzgPgLgngBwKYAIDCBDATjAkjJAWxQF4UIYsBLAOwHMVGUB6ZlBYDAYyQAsB7YABMkWFGH5iMKQpNRd+hDlS4YCQlPGQBYAFB6tqAKJL4AZRhrUZAHL98RCCzZD+NAOQwUAVwioiCPAowFQUAIQoAOqoAO4YNF4w/HpMqWnpGc7ikjwoXCFICU5JKLzxQsCoMLyhKBhOGCmZzZms5EgI2FZ59UgAdChGfXQDAERw/N6lGABuqDTJui3LaW1UBIROtCgT3mKqOKN6BoioAIJcMFRzFt1kAN4oAKo0nVRCDpsoIOjYeBvBUIwADcKAAvidkCgAAoYd4AMUktwIpBQj1h70+Th+mBwn0BFGBTRWmVhcEIhS8PxEXCohAwwAh+l0hl+OFIxJQUEGpjgKH4YB5gTgyKQnO5Fyuc35gsl1yQovFMLhQkRYgFyoRSMsBCAA===) the compiler indeed uses inheritance.

Does it fall short if you defined DU's recursively, like this?
```fsharp
type Tree =
    | Branch of int * Tree list
    | Leaf of int
```

**Reply by stroborobo** [reply]
Frankly, inheritance is an implementation detail, exhaustiveness however is a core property of sum types.

**Reply by keithn** [reply]
@stroborobo  you can actually promote the C# warning to an error if you want to get exhaustiveness checks, but it's not really a core property and I never miss it.

**Reply by stroborobo** [reply]
Could you do private abstract base and public sealed cases?

Don't get me wrong, I'd do a little happy dance when C# gets first class DU support, but the exhaustiveness just is part of its definition. And when you work with this property regularly and use it to prevent impossible states, it becomes a very helpful tool.

**Comment by keithn** [comment]
yeah, you can do recursion, so you can do something like...

```C#
public abstract record Tree
{
    public record Branch(int Value, ImmutableList<Tree> Children) : Tree;
    public record Leaf(int Value) : Tree;
}
```
